### ISSUE TITLE
Redisson Session Manager session attributes problem

### ISSUE THREAD

**Issue by DJakosa** [redisson/redisson#2476]
We want to use Redisson Session Manager as a Tomcat Session Manager. It is essential that Redisson is configured as a shared instance in Tomcat since plugins that need access to a shared Session will be used. We found a behavior that could potentially indicate a bug in the session management.

A simplified diagram of our system is shown in the following figure.

![download](https://user-images.githubusercontent.com/7360893/70802942-211b2500-1db3-11ea-8ea3-a30239ff5061.png)

The repository with the test project based on Spring Boot is here: https://github.com/DJakosa/redisson-test

**Expected behavior**
All apps should use the same shared Session which is resolved based on `JSESSIONID` cookie. The Session should contain attributes with latest values that also include session scoped beans.

**Actual behavior**
When the `readMode` property of Manager configuration in `context.xml` is set to `REDIS`, behavior seems to be as expected (see value of `session_attribute_count` in response or logs which is incremented by 1 for every request). But with `REDIS` read mode, session scoped beans are not stored as session attributes.

However, if we change readMode to `MEMORY`, session scoped beans are stored as session attributes. Behavior is as expected if requests are sent to apps with the same context path (app-a in Tomcat 1 and Tomcat 2). When the request is sent to a different app (e.g. to app-b after being sent to app-a before), the second app (app-b) gets session attributes with values of the last request to the first app (app-a). If the request is sent to the first app (app-a) again, the session attribute values such as `session_attribute_count` will not continue from the values that the second app (app-b) had set. It also seems that session attributes that are added to the Session by the second app (such as `session_attribute_app_b_value`) are not available in the first app.

**Steps to reproduce or test case**
1. Checkout repository with the test project
2. Build project with maven (`mvn clean verify`) to get war files at `./app-a/target/app-a.war` and `./app-b/target/app-b.war`
3. Run `docker-compose up` from project root to set up stack with two Tomcats and Redis
4. Send requests to following endpoints (use the same session identifier) and inspect response and logs:
   * http://localhost:8090/app-a/session-values (app-a.war in Tomcat 1)
   * http://localhost:8091/app-a/session-values (app-a.war in Tomcat 2)
   * http://localhost:8091/app-a2/session-values (app-a.war in Tomcat 2 at context /app-a2)
   * http://localhost:8091/app-b/session-values (app-b.war in Tomcat 2)
5. Set different `readMode` property value (`REDIS` or `MEMORY`), redeploy docker stack, flush Redis and repeat step 4.

**Redis version**
5.0.7

**Redisson version**
3.11.6

**Redisson configuration**
See `./volume/conf/redisson.yaml`

**Comment by spetar** [comment]
In our company we have the same issue: https://github.com/redisson/redisson/issues/2417

**Comment by ecolinet** [comment]
Same problem here !

Using `readMode=MEMORY` seems to fix the problem but given @DJakosa comments I don't feel confident...

And btw I think that `readMode=REDIS` (which is default mode) should also work.

**Comment by mrniko** [comment]
Thanks for the provided test project! It helped a lot to reproduce the issue. 

At the moment I see that `session_bean_count` is not incremented in REDIS mode.

**Comment by mrniko** [comment]
Spring doesn't call Session.setAttribute method on any bean method invocation. So it's not a Redisson bug. In MEMORY mode Session.getAttribute returns instance which might be changed but not stored in session through Session.setAttribute method. The same won't work in REDIS mode since it read value from Redis per Session.getAttribute invocation.

**Comment by mrniko** [comment]
> When the request is sent to a different app (e.g. to app-b after being sent to app-a before), the second app (app-b) gets session attributes with values of the last request to the first app (app-a). If the request is sent to the first app (app-a) again, the session attribute values such as session_attribute_count will not continue from the values that the second app (app-b) had set. It also seems that session attributes that are added to the Session by the second app (such as session_attribute_app_b_value) are not available in the first app.

Unable to reproduce this. since different contexts app-a2, app-b, app-a use own session id

**Comment by DJakosa** [comment]
Use the same session id for all apps to reproduce such behavior.

**Comment by mrniko** [comment]
appending JSESSIONID param in url doesn't override session id.

**Comment by DJakosa** [comment]
Set the value in JSESSIONID cookie.

**Comment by luiszimmermann** [comment]
Confirmed in my tests.
1. **Tomcat A** 9.0.36 with Redisson 3.12.5
2. **Tomcat B** 9.0.36 with Redisson 3.12.5
3. Redis 5.0.9 with 3 sentinels, 2 nodes

**context.xml** with `<Manager className="org.redisson.tomcat.RedissonSessionManager" configPath="${catalina.base}/conf/redisson.yaml" broadcastSessionEvents="true" readMode="REDIS" updateMode="DEFAULT" />`

With readMode="MEMORY" the flash messages session attributes works as expected.

**Comment by mrniko** [comment]
@luiszimmermann

session scoped beans should be used with @EnableRedissonHttpSession annotation. Did you set it?

**Comment by luiszimmermann** [comment]
> @luiszimmermann
> 
> session scoped beans should be used with @EnableRedissonHttpSession annotation. Did you set it?

Right now I am only using the redisson in tomcat, following [this strategy](https://github.com/redisson/redisson/blob/master/redisson-tomcat/README.md), without putting any redisson dependency in my app (grails 2.5.6/spring 4.2.9).

Should I add the redisson in my app and create the Configuration files with `EnableRedissonHttpSession` to make the `readMode="REDIS"` works with mutable session?

**Comment by mrniko** [comment]
Case you described covered by test org.redisson.tomcat.RedissonSessionManagerTest#testUpdateTwoServers. Could you point out there is an error in this test?

**Comment by alexbosy** [comment]
We have the same problem with our GRAILS 2.5.5 (based on Spring 4.1.8) application. As I understood GRAILS  flash messages and server side redirects are heavily using session scoped beans under the hood. So all this GRAILS features don't work, when using Redisson Session Manager with readMode="REDIS".
Also, It looks like this issue is related with #1868

**Comment by mrniko** [comment]
@alexbosy 

Can you provide some test case? Since my tests unable to reproduce the issue - org.redisson.tomcat.RedissonSessionManagerTest#testUpdateTwoServers

**Comment by alexbosy** [comment]
@mrniko As I mentioned above our problem is related with this #1868 issue. Grails 2.x "flash messages" or "chain() methods in controllers" are based on mutable session objects (the session object created and saved in the beginning of each request, can be changed till the end of the request processing ). After careful reading of #1868 issue I realised, that current redisson-tomcat version does not support mutable session objects in a right way at all, am I right?
If yes, I don't think you even need some legacy grails 2.x sample/test case to reproduce this problem. And PR proposed by @jchobantonov will fix all such issues.

**Comment by mrniko** [comment]
@alexbosy 

I attached new version which fixes the issue. Could you give it a try?

[redisson-tomcat-8-3.13.4-SNAPSHOT.jar.zip](https://github.com/redisson/redisson/files/5039853/redisson-tomcat-8-3.13.4-SNAPSHOT.jar.zip)

**Comment by alexbosy** [comment]
@mrniko I'll try it within next week. Thanks a lot!

**Comment by luiszimmermann** [comment]
@mrniko I'm using tomcat 9, can you provide this version? If yes, I will test asap.

**Comment by mrniko** [comment]
@luiszimmermann 

Here it is 
[redisson-tomcat-9-3.13.4-SNAPSHOT.jar.zip](https://github.com/redisson/redisson/files/5044781/redisson-tomcat-9-3.13.4-SNAPSHOT.jar.zip)

**Comment by mrniko** [comment]
@luiszimmermann 

Did you try it?

**Comment by alexbosy** [comment]
@mrniko I’ve tried this snapshot build with no luck.

**Our setup:** 

Standalone Tomcat 8.5.57
Redis server 5.0.7 with sentinel 
Zulu OpenJDK 1.8.0_242
Legacy Grails 2.5.5 application deployed as WAR file
redisson-all-3.13.3.jar and redisson-tomcat-8-3.13.4-SNAPSHOT.jar were copied to the tomcat lib folder

**RedissonSessionManager config in Tomcat context.xml:**

```
<Manager className="org.redisson.tomcat.RedissonSessionManager"
    configPath="${catalina.base}/conf/redisson-tomcat.json"
    readMode="REDIS" updateMode=“DEFAULT” broadcastSessionEvents="false"/>
```

**redisson-tomcat.json:**

```
{
   "sentinelServersConfig":{
   "idleConnectionTimeout":10000,
   "connectTimeout":3000,
   "timeout":2000,
   "retryAttempts":3,
   "retryInterval":1500,
   "failedSlaveReconnectionInterval":3000,
   "failedSlaveCheckInterval":60000,
   "password":null,
   "subscriptionsPerConnection":5,
   "clientName":null,
   "loadBalancer":{
   "class":"org.redisson.connection.balancer.RoundRobinLoadBalancer"
    },
    "subscriptionConnectionMinimumIdleSize":1,
    "subscriptionConnectionPoolSize":5,
    "slaveConnectionMinimumIdleSize":2,
    "slaveConnectionPoolSize":10,
    "masterConnectionMinimumIdleSize":2,
    "masterConnectionPoolSize":10,
    "readMode":"MASTER_SLAVE",
    "subscriptionMode":"SLAVE",
    "sentinelAddresses":[
    "redis://10.4.80.52:26379",
    "redis://10.4.80.53:26379",
    "redis://10.4.80.54:26379"
    ],
    "masterName":"master01",
    "database":2
    },
    "codec": {"class":"org.redisson.codec.MarshallingCodec"},
    "threads":16,
    "nettyThreads":32,
    "transportMode":"NIO"
}

```
**Got the following error (Probably some class loader related issue?):**

```
[redisson-netty-3-1] ERROR org.redisson.client.handler.ErrorsLoggingHandler - Exception occured. Channel: [id: 0x2c86536f, L:/10.4.80.62:48338 - R:10.4.80.54/10.4.80.54:6379]
io.netty.handler.codec.DecoderException: java.io.IOException: java.lang.ClassNotFoundException: org.codehaus.groovy.grails.web.servlet.GrailsFlashScope
	at io.netty.handler.codec.ReplayingDecoder.callDecode(ReplayingDecoder.java:421)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: java.lang.ClassNotFoundException: org.codehaus.groovy.grails.web.servlet.GrailsFlashScope
	at org.redisson.codec.MarshallingCodec$3.decode(MarshallingCodec.java:160)
	at org.redisson.client.handler.CommandDecoder.decode(CommandDecoder.java:378)
	at org.redisson.client.handler.CommandDecoder.decodeCommand(CommandDecoder.java:196)
	at org.redisson.client.handler.CommandDecoder.decode(CommandDecoder.java:134)
	at org.redisson.client.handler.CommandDecoder.decode(CommandDecoder.java:104)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:501)
	at io.netty.handler.codec.ReplayingDecoder.callDecode(ReplayingDecoder.java:366)
	... 17 more
Caused by: java.lang.ClassNotFoundException: org.codehaus.groovy.grails.web.servlet.GrailsFlashScope
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:352)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at org.jboss.marshalling.AbstractClassResolver.loadClass(AbstractClassResolver.java:129)
	at org.jboss.marshalling.AbstractClassResolver.resolveClass(AbstractClassResolver.java:110)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadClassDescriptor(RiverUnmarshaller.java:1033)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1366)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:283)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:216)
	at org.jboss.marshalling.AbstractObjectInput.readObject(AbstractObjectInput.java:41)
	at org.redisson.codec.MarshallingCodec$3.decode(MarshallingCodec.java:158)
	... 23 more
```
**Redis contents for the problematic session object:**

```
9) "org.codehaus.groovy.grails.FLASH_SCOPE"
10) "\x04\x04\t>7org.codehaus.groovy.grails.web.servlet.GrailsFlashScope\x14;\x0c:\b03<\x00\x00\x00\x02>\acurrent\x16\x00>\x04next\x16\x00\x16\x048>&java.util.concurrent.ConcurrentHashMapd\x99\xde\x12\x9d\x87)=\x00\x00\x00\x03>\x0bsegmentMask#\x00>\x0csegmentShift#\x00>\bsegments\x16\x00\x16\x00\x00\x00\x0f\x00\x00\x00\x1cB\x10\t>.java.util.concurrent.ConcurrentHashMap$Segment\x1f6L\x90X\x93)=\x00\x00\x00\x01>\nloadFactor&\x00\t>(java.util.concurrent.locks.ReentrantLockfU\xa8,,\xc8j\xeb\x00\x00\x00\x01>\x04sync\x16\x00\x16\x04;\xfe\x04\t>4java.util.concurrent.locks.ReentrantLock$NonfairSynce\x882\xe7S{\xbf\x0b\x00\x00\x00\x00\t>-java.util.concurrent.locks.ReentrantLock$Sync\xb8\x1e\xa2\x94\xaaDZ|\x00\x00\x00\x00\t>5java.util.concurrent.locks.AbstractQueuedSynchronizerfU\xa8Cu?R\xe3\x00\x00\x00\x01>\x05state#\x00\t>6java.util.concurrent.locks.AbstractOwnableSynchronizer3\xdf\xaf\xb9\xadmo\xa9\x00\x00\x00\x00\x16\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x01\x015\x04;\xf9\x00\x00\x00\x0f\x00\x00\x00\x1cB\x10;\xfa\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x04;\xfa\x04;\xfc\x00\x00\x00\x00?@\x00\x00\x01\x015"
```
Tried to changed the Codec in redisson-tomcat.json to org.redisson.codec.FstCodec.
There weren't any exceptions in this case, but GrailsFlashScope messages still didn’t work as expected. 

**Redis contents for the corresponding GrailsFlashScope object  were:**

```
15) "org.codehaus.groovy.grails.FLASH_SCOPE"
16) "\x00\x017org.codehaus.groovy.grails.web.servlet.GrailsFlashScope\x00\x11\x00\x00\x11\x00\x00"
```

I’ve also tried to change the updateMode=“DEFAULT” to updateMode=“AFTER_REQUEST” for both codecs.
I didn’t get any exceptions, but GrailsFlashScope messages still didn’t work as expected.

GrailsFlashScope is a mutable session object that includes other mutable objects -  several ConcurrentHashMaps. 
The contents of these maps can be changed during request processing multiple times.

The same approach (storing some state in mutable session objects) is heavily used in many other JVM web frameworks with server-side rendering  - 
Spring WebFlow, JSF, Wicket, Tapestry, etc

Here’s Grails 2.5.x GrailsFlashScope src:
https://github.com/grails/grails-core/blob/2.5.x/grails-web-common/src/main/groovy/org/codehaus/groovy/grails/web/servlet/GrailsFlashScope.java

**Comment by mrniko** [comment]
Here is update version. Please try it

[redisson-tomcat-8-3.13.4-SNAPSHOT.jar.zip](https://github.com/redisson/redisson/files/5102407/redisson-tomcat-8-3.13.4-SNAPSHOT.jar.zip)

**Comment by alexbosy** [comment]
> Here is update version. Please try it
> 
> [redisson-tomcat-8-3.13.4-SNAPSHOT.jar.zip](https://github.com/redisson/redisson/files/5102407/redisson-tomcat-8-3.13.4-SNAPSHOT.jar.zip)

it works!

**Comment by mrniko** [comment]
@alexbosy 

Thanks for help!

**Comment by alexbosy** [comment]
@mrniko But I tested it with only 1 tomcat for now. Will test it with at least 2 instances later.

**Comment by mrniko** [comment]
@alexbosy 

I see. Let me know once you test with 2 instances.

**Comment by mrniko** [comment]
@alexbosy 

Have you tried it with 2 instances yet?

**Comment by mrniko** [comment]
Fixed

**Comment by ghollies** [comment]
@mrniko this is breaking https://github.com/redisson/redisson/blob/master/redisson-tomcat/redisson-tomcat-8/src/test/java/org/redisson/tomcat/RedissonSessionManagerTest.java#L80
(test fails now, if i revert `b50bfcc` it passes)

It looks like `loadedAttributes` on server2 contains the stale value. for configurations other than context_memory.xml which still seems to work fine. 
```
org.junit.ComparisonFailure: 
Expected :324
Actual   :1234

	at org.junit.Assert.assertEquals(Assert.java:115)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at org.redisson.tomcat.RedissonSessionManagerTest.read(RedissonSessionManagerTest.java:256)
	at org.redisson.tomcat.RedissonSessionManagerTest.testUpdateTwoServers(RedissonSessionManagerTest.java:96)
```

I imagine this has broader implications beyond the broken test.

**Comment by mrniko** [comment]
@ghollies 

unable to reproduce it

**Comment by ghollies** [comment]
@mrniko I've created https://github.com/redisson/redisson/pull/3040 to show both the error I was talking about, and a related one.

the static field I was referring to was the `updateValve` field on the `RedissonSessionManager` during the creation of the second tomcat instance in the test, it sees that the static is set, and skips it, preventing either the UpdateValve or the UsageValve from being process for requests going to server2 in the tests.
That probably isn't an issue in the real world if the tomcat instances were their own JVMs, but it did help mask other errors that the tests can show.

I've added a few additional test cases to demonstrate the cases which fail
